test/e2e/backup_restore_test.go

```diff
@@ -0,0 +1,200 @@
+package e2e
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/iancoleman/strcase"
+	v1 "github.com/infinispan/infinispan-operator/pkg/apis/infinispan/v1"
+	v2 "github.com/infinispan/infinispan-operator/pkg/apis/infinispan/v2alpha1"
+	cconsts "github.com/infinispan/infinispan-operator/pkg/controller/constants"
+	ispnclient "github.com/infinispan/infinispan-operator/pkg/infinispan/client/http"
+	tconst "github.com/infinispan/infinispan-operator/test/e2e/constants"
+	tutils "github.com/infinispan/infinispan-operator/test/e2e/utils"
+	appsv1 "k8s.io/api/apps/v1"
+	k8errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type clusterSpec func(name, namespace string, clusterSize int) *v1.Infinispan
+
+func TestBackupRestore(t *testing.T) {
+	t.Run(string(v1.ServiceTypeDataGrid), testBackupRestore(datagridService))
+	t.Run(string(v1.ServiceTypeCache), testBackupRestore(cacheService))
+}
+
+func testBackupRestore(clusterSpec clusterSpec) func(*testing.T) {
+	return func(t *testing.T) {
+		// Create a resource without passing any config
+		name := strcase.ToKebab(strings.Replace(t.Name(), "/", "", 1))
+		namespace := tconst.Namespace
+		clusterSize := 2
+		numEntries := 100
+
+		// 1. Create initial source cluster
+		sourceCluster := name + "-source"
+		infinispan := clusterSpec(sourceCluster, namespace, clusterSize)
+		testKube.Create(infinispan)
+		defer testKube.DeleteInfinispan(infinispan, tconst.SinglePodTimeout)
+		waitForPodsOrFail(infinispan, clusterSize)
+
+		// 2. Populate the cluster with some data to backup
+		protocol := getSchemaForRest(infinispan)
+		cluster := newCluster(cconsts.DefaultOperatorUser, infinispan.GetSecretName(), protocol, testKube.Kubernetes)
+		cacheName := "someCache"
+		populateCache(cacheName, sourceCluster+"-0", numEntries, cluster.Client)
+
+		// 3. Backup the cluster's content
+		backupSpec := &v2.Backup{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+			Spec: v2.BackupSpec{
+				Cluster: sourceCluster,
+			},
+		}
+		testKube.Create(backupSpec)
+		defer testKube.DeleteBackup(backupSpec)
+
+		// Ensure the backup pod hased joined the cluster
+		waitForValidBackupPhase(name, namespace, v2.BackupSucceeded)
+
+		// Ensure that the backup pod has left the cluster, by checking a cluster pod's size
+		waitForPodsOrFail(infinispan, clusterSize)
+
+		// 4. Delete the original cluster
+		testKube.DeleteInfinispan(infinispan, tconst.SinglePodTimeout)
+		waitForNoCluster(sourceCluster)
+
+		// 5. Create a new cluster to restore the backup to
+		targetCluster := name + "-target"
```

Comment:
Recreate two nodes cluster is very time-consumed operation. Ii would prefer to use the same cluster for backup/restore just clean all records before restore.
@ryanemerson what do you think about it? Not sure that's very relates for testing
Also cache with name `default` always present in cluster and we can use it for testing.

Reply:
Restoring the backup to a new cluster is crucial imo as that's the main use-case for Backup/Restore, so I don't want to remove that aspect from the test.

I think we can remove the cache service tests after all. Long term we're removing the cache service and, other than having a defaullt cache defined, there shouldn't be any real difference betwen the two services from a backup/restore perspective.

> Also cache with name default always present in cluster and we can use it for testing.

Isn't that only true when using the cache service?